src/main.rs

```diff
@@ -243,7 +243,12 @@ mod tests {
         thread::sleep(Duration::from_millis(FIRECRACKER_INIT_TIMEOUT_MILLIS));
 
         // If Firecracker hasn't finished initializing yet, something is really wrong!
-        assert!(Path::new(DEFAULT_API_SOCK_PATH).exists());
+        if !Path::new(DEFAULT_API_SOCK_PATH).exists() {
+            // In this case we need to remove the API socket so other tests don't fail
+            // because we did not clean it up.
+            fs::remove_file(DEFAULT_API_SOCK_PATH).expect("failure in removing socket file");
```

Comment:
Why do you remove a file that `Path::new` couldn't find?

Reply:
LOL! That's not what the author wanted to do. So, the problem is that we are doing the assert on path here and it might not exist, but we do not kill the thread in which we spawned the Firecracker process, so if the test fails it will leave behind /tmp/firecracker.socket and `cargo test` for gnu and musl will fail with "Path already exists" which is confusing.

Maybe we should remove the default path on panic?

Reply:
But if the thread that's running `main` still isn't done with creating the socket, we can still successfully remove a non-existing file only for it to appear later, when `main` finally catches up. The only way to make sure the socket file isn't there is to make *sure* `main` is dead and only then remove it.
I suspect this happens only when you run `cargo test` (or `kcov`) outside the container? The container should deal with the cleanup.

Reply:
> But if the thread that's running `main` still isn't done with creating the socket, we can still successfully remove a non-existing file only for it to appear later, when `main` finally catches up. The only way to make sure the socket file isn't there is to make _sure_ `main` is dead and only then remove it.

Yes, I agree.

> I suspect this happens only when you run `cargo test` (or `kcov`) outside the container? The container should deal with the cleanup.


Neah, the container is not doing any cleanup. Here is the scenario that made this PR fail before:

First test to run is `test_coverage.py`. `test_main()` panics due to timeout on API socket. The Firecracker binary is still running and it is only killed when kcov finished the execution (this is a problem in itself). The file is not deleted because the assert made the test thread exit before it actually removed `/tmp/firecracker.socket`.

Second test to run is `test_unittest.py`. This test fails with "API path already exists" which is misleading. `test_unittests.py` should not fail, it only fails because we are not properly cleaning up the resources in the previous step.

Reply:
Aaah, it fails in the *same* test run :bulb: 
>Maybe we should remove the default path on panic?

I think the panic doesn't guarantee that the other thread (in this case, the one running `main`) is done, it needs to be `join`ed.

Reply:
If we run tests sequentially, can we forcefully kill the firecracker process at the end of a test?
For example in our naughty case, on test timeout `kill -9` the firecracker process from that test before moving on to the next.

Reply:
@acatangiu, yes we can, but that would end up killing the whole cargo test process. That shouldn't be a problem in itself if we decide to drop the `--no-fail-fast` parameter when running th unit tests. I've opened an issue for this particular problem: https://github.com/firecracker-microvm/firecracker/issues/1155